From df41834b71f0d0df3e164f98b79e735397dd2b43 Mon Sep 17 00:00:00 2001
From: Paolo Bonzini <pbonzini@redhat.com>
Date: Fri, 7 Nov 2014 15:55:56 -0500
Subject: [virt] kvm: emulate MOVNTDQ

Message-id: <1415375757-21342-1-git-send-email-pbonzini@redhat.com>
Patchwork-id: 99782
O-Subject: [RHEL7 PATCH] KVM: x86 emulator: emulate MOVNTDQ
Bugzilla: 1117542
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Laszlo Ersek <lersek@redhat.com>
RH-Acked-by: Alex Williamson <alex.williamson@redhat.com>

From: Alex Williamson <alex.williamson@redhat.com>

Bugzilla: 1117542

Windows 8.1 guest with NVIDIA driver and GPU fails to boot with an
emulation failure.  The KVM spew suggests the fault is with lack of
movntdq emulation (courtesy of Paolo):

Code=02 00 00 b8 08 00 00 00 f3 0f 6f 44 0a f0 f3 0f 6f 4c 0a e0 <66> 0f e7 41 f0 66 0f e7 49 e0 48 83 e9 40 f3 0f 6f 44 0a 10 f3 0f 6f 0c 0a 66 0f e7 41 10

$ as -o a.out
        .section .text
        .byte 0x66, 0x0f, 0xe7, 0x41, 0xf0
        .byte 0x66, 0x0f, 0xe7, 0x49, 0xe0
$ objdump -d a.out
    0:  66 0f e7 41 f0          movntdq %xmm0,-0x10(%rcx)
    5:  66 0f e7 49 e0          movntdq %xmm1,-0x20(%rcx)

Add the necessary emulation.

Signed-off-by: Alex Williamson <alex.williamson@redhat.com>
Cc: Paolo Bonzini <pbonzini@redhat.com>
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
(cherry picked from commit 0a37027e83f867793af0ccb9176a6b383dd0b7c8)
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/kvm/emulate.c b/arch/x86/kvm/emulate.c
index e4e833d..ae39f08 100644
--- a/arch/x86/kvm/emulate.c
+++ b/arch/x86/kvm/emulate.c
@@ -3681,6 +3681,10 @@ static const struct gprefix pfx_0f_28_0f_29 = {
  I(Aligned, em_mov), I(Aligned, em_mov), N, N,
 };
 
+static const struct gprefix pfx_0f_e7 = {
+ N, I(Sse, em_mov), N, N,
+};
+
 static const struct escape escape_d9 = { {
  N, N, N, N, N, N, N, I(DstMem, em_fnstcw),
 }, {
@@ -3951,7 +3955,8 @@ static const struct opcode twobyte_table[256] = {
  /* 0xD0 - 0xDF */
  N, N, N, N, N, N, N, N, N, N, N, N, N, N, N, N,
  /* 0xE0 - 0xEF */
- N, N, N, N, N, N, N, N, N, N, N, N, N, N, N, N,
+ N, N, N, N, N, N, N, GP(SrcReg | DstMem | ModRM | Mov, &pfx_0f_e7),
+ N, N, N, N, N, N, N, N,
  /* 0xF0 - 0xFF */
  N, N, N, N, N, N, N, N, N, N, N, N, N, N, N, N
 };
-- 
1.7.1